### PR TITLE
[SPARK-1292] In-memory columnar representation for Spark SQL

### PR DESCRIPTION
**Pull request by liancheng**
This PR is rebased from the Catalyst repository, and contains the first version of in-memory columnar representation for Spark SQL. Compression support is not included yet and will be added later in a separate PR.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlSerializer.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  override def newKryo(): Kryo = {
+    val kryo = new Kryo()
+    kryo.setRegistrationRequired(false)
+    kryo.register(classOf[MutablePair[_, _]])
+    kryo.register(classOf[Array[Any]])
+    kryo.register(classOf[org.apache.spark.sql.catalyst.expressions.GenericRow])
+    kryo.register(classOf[org.apache.spark.sql.catalyst.expressions.GenericMutableRow])
+    kryo.register(classOf[scala.collection.mutable.ArrayBuffer[_]])
+    kryo.register(classOf[scala.math.BigDecimal], new BigDecimalSerializer)
+    kryo.setReferences(false)
```
Comment (oliviertoupin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlSerializer.scala]: Why did you add setReferences to false? Is there a requirement for it? Why don't you use the value specified in the configuration?

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlSerializer.scala]: This is only used to shuffle SchemaRDDs and not user types.  Making `setReferences = false` improves performance.
